### ISSUE TITLE
Remapping keys

### ISSUE THREAD

**Issue by ioleo** [QubesOS/qubes-issues#5484]
**The problem you're addressing (if any)**
Currently I'm using a script (in my `.bash_profile`) like `setxkbmap pl -option ctrl:nocaps > /dev/null`, to remap CapsLock to Control. The script must be placed in each AppVM seperately and I have to make sure its run (eg. open console) before I can use the mapping.

**Describe the solution you'd like**
I'd like to be able to remap keyboard keys at QubesOS/dom0 level, such that the mapping is "transparent" to all AppVMs. 

**Where is the value to a user, and who might that user be?**
My use case is: being an Emacs user, I use Control a lot. However I don't remember a single moment last 10 year where I'd actually use CapsLock. It is convinient to remap it.

**Describe alternatives you've considered**
Current script serves me well, but it would be more convinient to define this mapping at top level.

**Additional context**
A more powerful version of this feature, allowing seperate mappings for each keyboard (with keyboard not being aware they're translated/mapped) and as early activated as the prompt for disc encryption password --- could perhaps serve as some kind of security for connecting external keyboards? Though I'm not a security expert, so just throwing in an idea, perhaps not good at all.

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

**Comment by strugee** [comment]
Does that command not work in dom0? I haven't tried it myself, I'm just curious.

If not you could also try working around this by putting the command in `/etc/profile` or something like that in your TemplateVMs. That might get started after X11 is started; I'm not sure. In that case you could try a systemd unit of `Type=oneshot` file ordered after `qubes-gui-agent.service`, again in the TemplateVM.

> could perhaps serve as some kind of security for connecting external keyboards?

Could you expand on why you think this would improve security? Is it because the keyboard wouldn't be able to log the "true" password since the user would be inputting the password _before_ it got mapped to the true password? If that's the case connecting an external (presumably USB) keyboard would still be a security risk because it could attack dom0's USB stack or execute commands in dom0 in a manner similar to the USB Rubber Ducky.

**Comment by ioleo** [comment]
@strugee yeah something along these lines... though as I stated, that was just a random thought
As for the script, yeah I could move it somewhere to be always run as soon as the AppVM is run, however (that would be a slight improvement over `.bash_profile` for which I have to open up the console at least once to activate the mapping), however as QubesOS is a single-user system I believe it would be nice to be able to define such mapping in one place for the whole system. Preferably that mapping would be transparent to the underlaying VMs.

Ideally this would be part of `Q -> System tools -> Keyboard` settings.

**Comment by marmarek** [comment]
> however as QubesOS is a single-user system I believe it would be nice to be able to define such mapping in one place for the whole system. Preferably that mapping would be transparent to the underlaying VMs.
> 
> Ideally this would be part of `Q -> System tools -> Keyboard` settings.

It is exactly the case. Keyboard layout (including options) is transferred from dom0 to all the VMs, as long as the dom0 keyboard layout is set sufficiently early. Changes of dom0 keyboard layout when VM is already running aren't transferred to that VM, yet (#1396).
But VM is free to override this layout setting - which is what your script do. Similarly per-vm keyboard layout tool available in qubes manager override this dom0-provided layout.

**Comment by ioleo** [comment]
So, indeed running `setxkbmap pl -option ctrl:nocaps` in dom0 does work in AppVMs! Great!
Though I've failed at makeing it autostart. I tried by adding `~/.config/autostart/RemapCapsLock.desktop` with contents:
```
[Desktop Entry]
Name=Remap Caps Lock
Comment=Remaps Caps Lock to Control
Exec=setxkbmap pl -option ctrl:nocaps
Terminal=true
StartupNotify=false
Type=Application
```

Restarted the machine, but it seems its not run.

**Comment by marmarek** [comment]
Have you tried to set this configuration simply in System Tools -> Keyboard ?

**Comment by marmarek** [comment]
Hmm, I don't see this specific option there...

As for the desktop file, try `Terminal=false`

**Comment by ioleo** [comment]
Tried `Terminal=false`, still not working. Perhaps it would be better to add as systemd service? I tried that too, but got into errors `cannot open display`, probably the script being run too early (I guess it needs X running first). This would probably be solved by the right `After=` setting, but at that point I stopped to try the `.desktop` autorstart approach.

**Comment by marmarek** [comment]
> got into errors `cannot open display`, probably the script being run too early (I guess it needs X running first)

That's exactly the problem with this idea - systemd services are run independent of graphical session. Keyboard layout needs to be set after X server startup, but also after each X server restart (logout).

What may be an issue here is a race with Xfce setting the layout - make sure you have "Use system defaults" option enabled in System Tools -> Keyboard.

**Comment by sysfu** [comment]
Any progress in adding this capslock key remapping to System Tools -> Keyboard?

Setting up Qubes for the first time today and not being able to remap the idiot key as CTRL is driving me crazy.

**Comment by ioleo** [comment]
No progress on my end. I actually ended up removing Qubes OS a while ago due to a heisenbug with all appVms suddenly failing to start (and as this is my working machine, I could not afford another week lost on backups and reinstalls). Although I hope to come back once I have another machine,

**Comment by moiselazarus** [comment]
I use a perfect tool for it: keyd; https://github.com/rvaiya/keyd. I remap capslock, space... there are endless possibilties: Layers, Key overloading ... The good thing, it works the same way in qube-vm terminals, but there are exeptions. I compile it in a fedora-qube and copy it to dom0.

**Comment by th5** [comment]
The following worked for me. I ran everything in Dom0 but changes also work in my AppVMs.

To try out the change without saving it permanently (rebooting clears it):
Dom0: ```setxkbmap -option ctrl:nocaps```

To make the change permanent:
Dom0: ```localectl set-x11-keymap "" "" "" ctrl:swapcaps```

localectl above will write to:
Dom0: ```cat /etc/X11/xorg.conf.d/00-keyboard.conf```

Checking current options:
```localectl status```

I'm not too familiar with localectl so perhaps there is a better way to use it. I wasn't sure about set-x11-keymap vs set-keymap.
